processing/src/main/java/io/druid/segment/QueryableIndexIndexableAdapter.java

```diff
@@ -339,79 +336,23 @@ public ColumnCapabilities getCapabilities(String column)
     }
 
-    final Indexed<String> dimSet = getDimValueLookup(dimension);
-
-    // BitmapIndexSeeker is the main performance boost comes from.
-    // In the previous version of index merge, during the creation of invert index, we do something like
-    // merge sort of multiply bitmap indexes. It simply iterator all the previous sorted values,
-    // and "binary find" the id in each bitmap indexes,  which involves disk IO and is really slow.
-    // Suppose we have N (which is 100 in our test) small segments,  each have M (which is 50000 in our case) rows.
-    // In high cardinality scenario, we will almost have N * M uniq values. So the complexity will be O(N * M * M * LOG(M)).
-
-    // There are 2 properties we did not use during the merging:
-    // 1. We always travel the dimension values sequentially
-    // 2. One single dimension value is valid only in one index when cardinality is high enough
-    // So we introduced the BitmapIndexSeeker, which can only seek value sequentially and can never seek back.
-    // By using this and the help of "getDimValueLookup", we only need to translate all dimension value to its ID once,
-    // and the translation is done by self-increase of the integer. We only need to change the CACHED value once after
-    // previous value is hit, renew the value and increase the ID. The complexity now is O(N * M * LOG(M)).
-    return new BitmapIndexSeeker()
-    {
-      private int currIndex = 0;
-      private String currVal = null;
-      private String lastVal = null;
-
-      @Override
-      public IndexedInts seek(String value)
-      {
-        if (dimSet == null || dimSet.size() == 0) {
-          return new EmptyIndexedInts();
-        }
-        if (lastVal != null) {
-          if (GenericIndexed.STRING_STRATEGY.compare(value, lastVal) <= 0) {
-            throw new ISE(
-                "Value[%s] is less than the last value[%s] I have, cannot be.",
-                value, lastVal
-            );
-          }
-          return new EmptyIndexedInts();
-        }
-        if (currVal == null) {
-          currVal = dimSet.get(currIndex);
-        }
-        int compareResult = GenericIndexed.STRING_STRATEGY.compare(currVal, value);
-        if (compareResult == 0) {
-          IndexedInts ret = new BitmapCompressedIndexedInts(bitmaps.getBitmap(currIndex));
-          ++currIndex;
-          if (currIndex == dimSet.size()) {
-            lastVal = value;
-          } else {
-            currVal = dimSet.get(currIndex);
-          }
-          return ret;
-        } else if (compareResult < 0) {
-          throw new ISE(
-              "Skipped currValue[%s], currIndex[%,d]; incoming value[%s]",
-              currVal, currIndex, value
-          );
-        } else {
-          return new EmptyIndexedInts();
-        }
-      }
-    };
+    if (dictId >= 0) {
+      return new BitmapCompressedIndexedInts(bitmaps.getBitmap(dictId));
+    } else {
+      return new EmptyIndexedInts();
```

Comment:
I believe EmptyIndexedInts should be a singleton, it already has a static final instance, can you use that? also make the no arg constructor in that class be private.